enhancements/proposals/1978-scylladbdatacenter-api/README.md

```diff
@@ -0,0 +1,724 @@
+
+#### Node count
+
+A new field `nodes` will be added on the datacenter level. Scylla Operator will balance this number between all defined racks.
+Having a single place where the desired number of nodes is specified, enables us to have support for scale subresource,
+which brings users handy scaling operations using `kubectl scale` and paves a path towards available Kubernetes autoscalers.
+The downside of it is that these won't be able to add more racks, but provides a nice way of scaling for static topology clusters.
+
+Datacenter level `nodes` cannot be used together with `nodes` specified per rack. Webhook will prevent it.
+
+The balancing algorithm is iterative, adding/removing one node at a time in each iteration to ensure that racks stay as balanced as possible.
+
+```go
+func balanceNodesAcrossRacks(racks []*appsv1.StatefulSet, totalNodes int) []*appsv1.StatefulSet {
+    // Calculate the total number of existing nodes
+    existingNodes := 0
+    for _, rack := range racks {
+       existingNodes += rack.Replicas
+    }
+
+    if existingNodes == totalNodes {
+       // The desired number of nodes reached, do nothing.
+       return racks
+    }
+    
+    if existingNodes < totalNodes {
+       // Add a node to the least populated rack
+       minRackIndex := findSmallestStatefulSet(racks)
+       racks[minRackIndex].Replicas++
+       return balanceNodesAcrossRacks(racks, totalNodes)
+    }
+    
+    // Remove a node from biggest rack.
+    maxRackIndex := findBiggestStatefulSet(racks)
+    racks[maxRackIndex].Replicas--
+    return balanceNodesAcrossRacks(racks, totalNodes)
+}
+```
+
+New Condition will be added to the `status` of `v1alpha1.ScyllaDBDatacenter`, providing a simple way of noticing rack imbalance.
+Its type is going to be `UnbalancedRacks` and the value of it will be reconciled accordingly to the state of rack imbalance.
+
+--- 
+
+An example of nine node cluster having three racks with uniform sizes and resources.
+```yaml
+scylladb:
+  image: docker.io/scylladb/scylla:6.0.1
+  resources:
+    requests:
+      cpu: 1
+      memory: 4Gi
+      storage: 100Gi
+scyllaDBManagerAgent:
+  image: docker.io/scylladb/scylla-manager-agent:3.3.0
+  resources:
+    requests:
+      cpu: 100m
+      memory: 100Mi
+nodes: 9
+racks:
+- name: a
+- name: b
+- name: c
+
+```
+
+### API Spec
+
+Proposed API spec:
+```
+
+const (
+    AvailableCondition   = "Available"
+    ProgressingCondition = "Progressing"
+    DegradedCondition    = "Degraded"
+)
+
+// ScyllaDBDatacenterSpec defines the desired state of ScyllaDBDatacenter.
+type ScyllaDBDatacenterSpec struct {
+    // metadata controls shared metadata for all pods created based on this spec.
+    // +optional
+    Metadata *ObjectTemplateMetadata `json:"metadata,omitempty"`
+
+    // datacenterName is the name of the scylla datacenter. Used as datacenter name in GossipingPropertyFileSnitch.
+    DatacenterName string `json:"datacenterName"`
```

Comment:
Imo yes, otherwise it would be confusing as datacenter name is propagated into scylla and visible to the user. In such case, you couldn't have two `us-east1` datacenters in the same namespace, even though they could be part of different cluster.

But it raised an important issue we need to solve as well. There's also a `cluster_name` property of Scylla, which should be in sync between datacenters in order for two dcs to join each other. So far, we used `scyllacluster.metadata.name` for it, but since we changed the name of CRD, it doesn't make sense anymore. 

Maybe we should have a dedicated field for all three (cluster, dc, rack), wdyt?

Reply:
I think this makes sense. Making them optional with defaults based on the metadata.name for the simple cases would make a nice UX.

Reply:
done